log: check fclose() results and report safely in logging.cpp

cedwies opened the pull request:
`fclose()` can report write errors (for example if the disk is full or the filesystem has a problem). Right now, some `fclose()` calls in `src/logging.cpp` ignore the return value. This means errors might go unnoticed and log lines could be lost without warning.

## What this PR does:
- Add a small helper that prints `fclose()` errors to `stderr` (with path and errno).
- In shutdown: close `m_fileout` safely and report errors.
- In reopen: open the new file, swap it in, close the old one, and report errors if closing fails.
- In shrink/rotate: check all `fclose(file)` calls and report failures.

No other behavior changes. Normal logging, rotation, and console output remain unchanged.

cedwies committed:
log: check fclose() results and report safely

Some buffered I/O errors surface only on fclose(). Several fclose() calls in
src/logging.cpp ignored the return value which can risk silent loss of log lines
or misleading success during rotate/truncate.

Check return of fclose() and include errno in diagnostics. For the active log sink,
detach under lock and close outside the lock. On failure, report directly
to stderr (which avoids recursive logging) and flush. No behavior change on success.

ajtowns reviewed (commented):
Should rebase on top of master rather than including a merge commit.

sedited commented:
This looks ok, but is there a way to consistently trigger an fclose error so this might be tested locally?

sedited commented:
@cedwies ping for rebase.

cedwies commented:
Rebased on current master.

I don’t see an easy clean unit test for this without adding some kind of FILE*/fs fault injection. Since the log file is unbuffered, normal write errors should usually show up on fwrite, as @maflcko said. So I think this is mainly a small defensive check for close/writeback errors.

maflcko commented on src/logging.cpp:
```diff
@@ -16,13 +18,29 @@
+// Report a log I/O failure without re-entering the logger.
+// Used when opening/closing the active log file. We intentionally write to
+// stderr (and not Log*), because this may be called while the logger mutex
+// is held or while the log file is failing. Do not call Log* from here.
```
The comment should be on the function, not the namespace?

cedwies replied:
Done, the documentation is now attached directly to `CloseLogFile()`.

sedited commented on src/logging.cpp:
```diff
@@ -97,15 +115,20 @@ bool BCLog::Logger::StartLogging()
+        m_msgs_before_open.clear();
+    }
+    if (f && fclose(f) != 0) {
+        ReportLogIOError("fclose", m_file_path);
```
Is the idea that `ReportLogIOError` is never called under a lock? Reading through the other calls to the function, I do see some scenarios where the lock is being held. So changing this for this single test-only path seems like a poor motivation.

cedwies replied:
Agreed. I restored the close in `DisconnectTestLogger()` to run under `m_cs` for consistency with the other call sites.

sedited commented on src/logging.cpp:
```diff
@@ -540,17 +566,24 @@ void BCLog::Logger::ShrinkDebugFile()
         }
         int nBytes = fread(vch.data(), 1, vch.size(), file);
-        fclose(file);
+        if (fclose(file) != 0){
```
Nit (clang-format): There are some whitespace issues throughout this change. Can you run clang-format over the patch? I usually use the following command (and then only selectively apply the patches where they make sense): `git diff -U0 HEAD~1.. | ./contrib/devtools/clang-format-diff.py -p1 -i -v`.

cedwies replied:
I ran clang-format using the suggested command.

sedited commented:
@cedwies there's been some pending review comments here for a while. Do you have time to get around to them?

cedwies commented:
Rebased onto current master and addressed the pending review comments:
1. moved the helper documentation onto the function;
2. kept the `DisconnectTestLogger()` close under `m_cs`
3. ran clang-format as suggested.

I also consolidated the close/error reporting sequence into `CloseLogFile()`, so now all `fclose()` calls in `logging.cpp` are checked consistently  (including the `fseek()` failure path). The duplicated numeric errno was removed, too.
@sedited @maflcko, re-review would be appreciated when convenient, thank you for your patience.
